File: starter_kits/Ruby/hlt/position.rb
require 'direction'

class Position
  attr_reader :x, :y

  def initialize(x, y)
    @x = x
    @y = y
  end

  # Returns the position considering a Direction cardinal tuple
  # :param direction: the direction cardinal array
  # :return: a new position moved in that direction
  def directional_offset(direction)
    return self + Position.new(*direction)
  end

  # :return: Returns a list of all positions around this specific position in
  #   each cardinal direction
  def get_surrounding_cardinals
    Direction.all_cardinals.map do |direction|
      directional_offset(direction)
    end
  end

  def +(other)
    Position.new(@x + other.x, @y + other.y)
  end

  def -(other)
    Position.new(@x - other.x, @y - other.y)
  end

  def abs
    Position.new(@x.abs, @y.abs)
  end

  def ==(other)
    return other.nil? ? false : @x == other.x && @y == other.y
  end

  def to_s
    "Position (#{@x}, #{@y})"
  end
end

File: starter_kits/TypeScript/hlt/Position.ts
import { Direction } from "./Direction";

export class Position {
    constructor(public x: number, public y: number) {
        this.x = x;
        this.y = y;
    }

    public directionalOffset(direction: Direction) {
        return this.add(new Position(direction.dx, direction.dy));
    }

    public getSurroundingCardinals() {
        return Direction.getAllCardinals()
            .map((currentDirection) => this.directionalOffset(currentDirection));
    }

    public add(other: Position) {
        return new Position(this.x + other.x, this.y + other.y);
    }

    public sub(other: Position) {
        return new Position(this.x - other.x, this.y - other.y);
    }

    public addMut(other: Position) {
        this.x += other.x;
        this.y += other.y;
    }

    public subMut(other: Position) {
        this.x -= other.x;
        this.y -= other.y;
    }

    public abs() {
        return new Position(Math.abs(this.x), Math.abs(this.y));
    }

    public equals(other: Position) {
        return this.x === other.x && this.y === other.y;
    }

    public toString() {
        return `${this.constructor.name}(${this.x}, ${this.y})`;
    }
}
